homeassistant/helpers/device_registry.py

```diff
@@ -210,13 +188,76 @@ async def _async_migrate_func(
         return old_data
 
 
+_EntryTypeT = TypeVar("_EntryTypeT", DeviceEntry, DeletedDeviceEntry)
+
+
+class DeviceRegistryItems(UserDict[str, _EntryTypeT]):
+    """Container for device registry items, maps device id -> entry.
+
+    Maintains two additional indexes:
+    - (connection_type, connection identifier) -> entry
+    - (DOMAIN, identifier) -> entry
+    """
+
+    def __init__(self) -> None:
+        """Initialize the container."""
+        super().__init__()
+        self._connections: dict[tuple[str, str], _EntryTypeT] = {}
+        self._identifiers: dict[tuple[str, str], _EntryTypeT] = {}
+
+    def __setitem__(self, key: str, entry: _EntryTypeT) -> None:
+        """Add an item."""
+        if key in self:
+            old_entry = self[key]
+            for connection in old_entry.connections:
+                del self._connections[connection]
+            for identifier in old_entry.identifiers:
+                del self._identifiers[identifier]
+        super().__setitem__(key, entry)  # type: ignore[assignment]
```

Comment:
Maybe add a comment referencing https://github.com/python/mypy/issues/13596
```suggestion
       # type ignore linked to mypy issue: https://github.com/python/mypy/issues/13596
        super().__setitem__(key, entry)  # type: ignore[assignment]
```